driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +191,154 @@ public int getMaxWireVersion() {
         return maxWireVersion;
     }
 
-    private boolean limitReached() {
-        return Math.abs(limit) != 0 && count.get() >= Math.abs(limit);
+    private void getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
+        resourceManager.executeWithConnection((connection, wrappedCallback) ->
+                        assertNotNull(connection).commandAsync(namespace.getDatabaseName(),
+                                getMoreCommandDocument(cursor.getId(), connection.getDescription(), namespace,
+                                        limit, batchSize, count.get(), maxTimeMS, comment),
+                                NO_OP_FIELD_NAME_VALIDATOR, ReadPreference.primary(),
+                                CommandResultDocumentCodec.create(decoder, NEXT_BATCH),
+                                assertNotNull(resourceManager.getConnectionSource()).getOperationContext(),
+                                (commandResult, t) -> {
+                                    if (t != null) {
+                                        Throwable translatedException =
+                                                t instanceof MongoCommandException
+                                                        ? translateCommandException((MongoCommandException) t, cursor)
+                                                        : t;
+                                        wrappedCallback.onResult(null, translatedException);
+                                        return;
+                                    }
+                                    CommandCursorResult<T> commandCursorResult = toCommandCursorResult(
+                                            connection.getDescription().getServerAddress(), NEXT_BATCH, assertNotNull(commandResult));
+                                    resourceManager.setServerCursor(commandCursorResult.getServerCursor());
+
+                                    if (!resourceManager.operable()) {
+                                        // The cursor was closed
+                                        resourceManager.releaseServerAndClientResources(connection);
+                                        wrappedCallback.onResult(emptyList(), null);
+                                        return;
+                                    }
+
+                                    List<T> nextBatch = commandCursorResult.getResults();
+                                    if (nextBatch.isEmpty() && commandCursorResult.getServerCursor() != null) {
+                                        getMore(commandCursorResult.getServerCursor(), wrappedCallback);
```

Comment:
I think this will check out another connection before returning the current one to the pool?                                        // Should it be re-using the same connection(was it doing that previously?)?

Reply:
Fixed